### Calculating the centrality of every node in the Arpanet

In [27]:
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

Here is an image of the Arpanet, according to homework assignment 1: 
<img src='https://user-images.githubusercontent.com/33112694/69269469-fcbf9480-0bc8-11ea-92a3-73fb4f689ec7.png' />

For edges without weight values, we assume that the weight is 0.1, 
and here is the corresponding adjacent matrix:

|&nbsp;| RAND | UCSB | SU   | UCB | SRI | UCLA | SDC | UTAH | WU   | ILL  | MICH | CMU  | ARPA | BTL  | HARV | LL   | BBN  | MAC  | DART |
|------|------|------|------|-----|-----|------|-----|------|------|------|------|------|------|------|------|------|------|------|------|
| RAND | 0    | 0\.3 | 0    | 0   | 0   | 0    | 1   | 2    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| UCSB | 0\.3 | 0    | 0\.7 | 0   | 0   | 0\.3 | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| SU   | 0    | 0\.7 | 0    | 1   | 0   | 0    | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 8    | 0    |
| UCB  | 0    | 0    | 1    | 0   | 1   | 0    | 0   | 1    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| SRI  | 0    | 0    | 0    | 1   | 0   | 1    | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 8    | 0    | 0    | 0    | 0    |
| UCLA | 0    | 0\.3 | 0    | 0   | 1   | 0    | 1   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| SDC  | 1    | 0    | 0    | 0   | 0   | 1    | 0   | 0    | 5    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| UTAH | 2    | 0    | 0    | 1   | 0   | 0    | 0   | 0    | 0    | 4    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| WU   | 0    | 0    | 0    | 0   | 0   | 0    | 5   | 0    | 0    | 0\.4 | 0    | 1\.8 | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| ILL  | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 4    | 0\.4 | 0    | 1    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    |
| MICH | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 1    | 0    | 0    | 0    | 1\.5 | 0    | 0    | 0    | 0    | 1\.8 |
| CMU  | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 1\.8 | 0    | 0    | 0    | 0\.5 | 0    | 0    | 0    | 1\.4 | 0    | 0    |
| ARPA | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 0    | 0\.5 | 0    | 0\.6 | 0    | 0    | 0    | 0    | 0    |
| BTL  | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 1\.5 | 0    | 0\.6 | 0    | 0    | 0\.6 | 0    | 0    | 0    |
| HARV | 0    | 0    | 0    | 0   | 8   | 0    | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0\.1 | 0    | 0    | 0\.3 |
| LL   | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0\.6 | 0\.1 | 0    | 0\.1 | 0    | 0    |
| BBN  | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 0    | 1\.4 | 0    | 0    | 0    | 0\.1 | 0    | 0\.1 | 0    |
| MAC  | 0    | 0    | 8    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0\.1 | 0    | 0\.3 |
| DART | 0    | 0    | 0    | 0   | 0   | 0    | 0   | 0    | 0    | 0    | 1\.8 | 0    | 0    | 0    | 0\.3 | 0    | 0    | 0\.3 | 0    |

In [28]:
adj_mat = [
    [0,0.3,0,0,0,0,1,2,0,0,0,0,0,0,0,0,0,0,0],      # RAND
    [0.3,0,0.7,0,0,0.3,0,0,0,0,0,0,0,0,0,0,0,0,0],  # UCSB
    [0,0.7,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,8,0],      # SU
    [0,0,1,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0],        # UCB
    [0,0,0,1,0,1,0,0,0,0,0,0,0,0,8,0,0,0,0],        # SRI
    [0,0.3,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0],      # UCLA
    [1,0,0,0,0,1,0,0,5,0,0,0,0,0,0,0,0,0,0],        # SDC
    [2,0,0,1,0,0,0,0,0,4,0,0,0,0,0,0,0,0,0],        # UTAH
    [0,0,0,0,0,0,5,0,0,0.4,0,1.8,0,0,0,0,0,0,0],    # WU
    [0,0,0,0,0,0,0,4,0.4,0,1,0,0,0,0,0,0,0,0],      # ILL
    [0,0,0,0,0,0,0,0,0,1,0,0,0,1.5,0,0,0,0,1.8],    # MICH
    [0,0,0,0,0,0,0,0,1.8,0,0,0,0.5,0,0,0,1.4,0,0],  # CMU
    [0,0,0,0,0,0,0,0,0,0,0,0.5,0,0.6,0,0,0,0,0],    # ARPA
    [0,0,0,0,0,0,0,0,0,0,1.5,0,0.6,0,0,0.6,0,0,0],  # BTL
    [0,0,0,0,8,0,0,0,0,0,0,0,0,0,0,0.1,0,0,0.3],    # HARV
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0.6,0.1,0,0.1,0,0],  # LL
    [0,0,0,0,0,0,0,0,0,0,0,1.4,0,0,0,0.1,0,0.1,0],  # BBN
    [0,0,8,0,0,0,0,0,0,0,0,0,0,0,0,0,0.1,0,0.3],    # MAC
    [0,0,0,0,0,0,0,0,0,0,1.8,0,0,0,0.3,0,0,0.3,0]   # DART
]

#### Degree centrality
According to the formula, we have

<img src='https://user-images.githubusercontent.com/33112694/66724949-b170d400-ee1b-11e9-96f7-357996eadb8a.png' />

In [29]:
def degree_centrality(mat, weighted=False, alpha=1):
    n = len(mat)
    if n == 0:
        return None
    
    result = []
    for row in mat:
        if not weighted:
            nw_score = sum([1 for x in row if x > 0])
            result.append(nw_score / (n - 1))
        else:
            nw_score = sum([1 for x in row if x > 0])
            w_score = sum(row)
            result.append(nw_score ** (1 - alpha) * w_score ** alpha / (n - 1))
            
    return result

d_cent_non_weighted = degree_centrality(adj_mat, weighted=False)
print(d_cent_non_weighted)
d_cent_weighted = degree_centrality(adj_mat, weighted=True, alpha=0.5)
print(d_cent_weighted)

[0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.1111111111111111, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666, 0.16666666666666666]
[0.17480147469502524, 0.10971343143406388, 0.2996911990946928, 0.16666666666666663, 0.3042903097250923, 0.14593250596181884, 0.2545875386086578, 0.2545875386086578, 0.25819888974716115, 0.22360679774997896, 0.19953649995631076, 0.18509256943286312, 0.08240220541217405, 0.15811388300841897, 0.27888667551135854, 0.08606629658238703, 0.1217161238900369, 0.27888667551135854, 0.14907119849998599]


If we ignore edge weights, the degree centrality for each node is:

| RAND | UCSB | SU   | UCB | SRI | UCLA | SDC | UTAH | WU   | ILL  | MICH | CMU  | ARPA | BTL  | HARV | LL   | BBN  | MAC  | DART |
|------|------|------|-----|-----|------|-----|------|------|------|------|------|------|------|------|------|------|------|------|
| 0.167| 0.167| 0.167|0.167|0.167| 0.167|0.167| 0.167| 0.167| 0.167| 0.167| 0.167| 0.111| 0.167| 0.167| 0.167| 0.167| 0.167| 0.167|

However, according to <i>Node Centrality in Weighted Networks: Generalizing Degree and Shortest Paths</i>
by Tore Opsahl et al., we could improve the algorithm by combine the weighted node strength and the non-weighted node strength.

Here is the weighted degree centrality:

| RAND | UCSB | SU   | UCB | SRI | UCLA | SDC | UTAH | WU   | ILL  | MICH | CMU  | ARPA | BTL  | HARV | LL   | BBN  | MAC  | DART |
|------|------|------|-----|-----|------|-----|------|------|------|------|------|------|------|------|------|------|------|------|
| 0.175|0.110 | 0.300|0.167|0.304| 0.146|0.255| 0.255| 0.258| 0.223| 0.200| 0.185| 0.082| 0.158| 0.279| 0.086| 0.122| 0.279| 0.149|
